Improved memory management code in ParseSoaReply()

uttampawar opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
Test results:
$ python tools/test.py -I 'parallel/test-dns*'
 [00:00|% 100|+  12|-   0]: Done

$ make test
[06:23|% 100|+ 2429|-   2]: Done
Failed tests are not due to this change.

- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

This change was to improve memory management by use of MallocedBuffer instead of malloc()/free() calls. I further removed use of structure "ares_soa_reply" by using just local variables which removed 2 additional malloc/free calls via ares_expand_name().

cjihrig commented on src/cares_wrap.cc:
```diff
@@ -1064,30 +1064,29 @@ int ParseSoaReply(Environment* env,
   unsigned char* ptr = buf + NS_HFIXEDSZ;
-  char* name;
-  char* rr_name;
+  int rr_type, rr_len;
```
Should this line be here?

BridgeAR replied:
It does seem to be added by accident?

uttampawar replied:
@cjihrig @BridgeAR Do you mean line# 1067?

cjihrig replied:
Yes. It looks related to merge conflicts or something (I'm pretty sure I removed that line recently).

uttampawar replied:
@cjihrig Yes you are right, that line is not in the mainline. Is it okay if I update the PR with removal of that line (rr_len and rr_type)?

Trott commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/17845/

refack commented on src/cares_wrap.cc:
```diff
@@ -1064,30 +1064,28 @@ int ParseSoaReply(Environment* env,
 
   for (unsigned int i = 0; i < ancount; i++) {
-    status = ares_expand_name(ptr, buf, len, &rr_name, &temp_len);
+    status = ares_expand_name(ptr, buf, len, &rr_name.data, &temp_len);
```
[ES.26](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-recycle): Don’t use a variable for two unrelated purposes

refack commented on src/cares_wrap.cc:
```diff
@@ -1098,76 +1096,67 @@ int ParseSoaReply(Environment* env,
     /* only need SOA */
     if (rr_type == ns_t_soa) {
-      ares_soa_reply soa;
+      MallocedBuffer<char> nsname;
```
ditto

refack commented on src/cares_wrap.cc:
```diff
@@ -1098,76 +1096,67 @@ int ParseSoaReply(Environment* env,
+      MallocedBuffer<char> hostmaster;
 
-      status = ares_expand_name(ptr, buf, len, &soa.nsname, &temp_len);
+      status = ares_expand_name(ptr, buf, len, &nsname.data, &temp_len);
```
"

refack commented on src/cares_wrap.cc:
```diff
@@ -1098,76 +1096,67 @@ int ParseSoaReply(Environment* env,
       ptr += temp_len;
 
-      status = ares_expand_name(ptr, buf, len, &soa.hostmaster, &temp_len);
+      status = ares_expand_name(ptr, buf, len, &hostmaster.data, &temp_len);
```
"

refack commented:
Hello @uttampawar and thank you for the contribution.
I've left some comments on how to get the code be conform to our guidelines. I'm happy to help with anything you need.

uttampawar commented:
@refack Thanks for the feedback. I did read up on C++ core guidelines as you suggested and it make sense to use unique_ptr interface for small usages like above. Let me come back with suggestion/modification.

uttampawar commented:
@addaleax @refack I like the solution suggested by @refack to use unique_ptr. At the same time, I could be wrong but I believe MallocedBuffer was introduced to create a common interface for all memory management patterns especially malloc() and free(). Also, IMHO in this case MallocedBuffer destructor should be using cares API (ares_free_string()) to free the memory instead of free(). Thanks to @refack for pointing that out.  
IMHO, suggested solution is okay to use just in one place, but could it cause code bloat if it has to be repeated at multiple places? Looking for additional feedback.

refack commented:
@uttampawar thank you for following up!

> I believe MallocedBuffer was introduced to create a common interface for all memory management patterns especially malloc() and free().

IMHO `MallocedBuffer` has many uses, and there are cases where it fits 100%, but I don't think it should be used for all memory management.
I think this case a `unique_ptr` specialization fits better.

As for code bloat, you could define:
```cpp
struct AresDeleter {
  void operator()(char* ptr) const { ares_free_string(ptr); } noexcept
};
using ares_unique_ptr = std::unique_ptr<char, AresDeleter>;
```
in `cares_wrap.h`. That way all `cares` related memory management and ownership taking code could reuse it.

addaleax commented:
(Fwiw, in this case it would be `std::unique_ptr<char[]>`, not just `char` – we’ve apparently had trouble with `shared_ptr` on this point in the past, but for `unique_ptr` it should work.)

refack commented on src/cares_wrap.cc:
```diff
@@ -1064,30 +1064,28 @@ int ParseSoaReply(Environment* env,
   unsigned char* ptr = buf + NS_HFIXEDSZ;
-  char* name;
-  char* rr_name;
+  MallocedBuffer<char> name;
```
Since there is no further transfer of ownership, I think that is you use this pattern, i.e.:
```cpp
long temp_len;  // NOLINT(runtime/int)
char* name_tmp;
int status = ares_expand_name(ptr, buf, len, &name_tmp, &temp_len);
const ares_unique_ptr name(name_tmp);
```
You can use a `const ares_unique_ptr` getting more explicit code, and more compiler validation.
From [CppReference](https://en.cppreference.com/w/cpp/memory/unique_ptr):
> Only non-const unique_ptr can transfer the ownership of the managed object to another unique_ptr. If an object's lifetime is managed by a const std::unique_ptr, it is limited to the scope in which the pointer was created.

uttampawar commented:
@addaleax @refack It looks like we have an agreement on using unique_ptr instead of MallocedBuffer API for this case. I agree with suggestion of putting 'AresDeleter' in appropriate header file. I'll update the PR.

uttampawar commented:
@refack @addaleax I've updated my PR with new changes as discussed, using std::unique_ptr instead of MallocedBuffer<>.

refack commented on src/cares_wrap.cc:
```diff
@@ -1061,33 +1061,39 @@ int ParseSoaReply(Environment* env,
   }
   ptr += temp_len + NS_QFIXEDSZ;
 
+  char* rr_name_temp;
```
could you move this inside the `for`? [ES.5](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-scope): Keep scopes small
also please declare a new `long temp_rr_len` and `int status2` [ES.26](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-recycle): Don’t use a variable for two unrelated purposes

Reusing variables has no runtime benefit, and has the risk of misuse.

refack commented on src/cares_wrap.cc:
```diff
@@ -1096,78 +1102,71 @@ int ParseSoaReply(Environment* env,
+      char* hostmaster_temp;
 
-      status = ares_expand_name(ptr, buf, len, &soa.nsname, &temp_len);
+      status = ares_expand_name(ptr, buf, len, &nsname_temp, &temp_len);
```
Same here

uttampawar replied:
@refack Currently the status to return is captured in one variable 'status'. In a for loop, all the 'break' statements breaks out of loop and return some status. Having multiple of these small scoped variables breaking the flow by having multiple return statements instead of break in a loop.  I've not seen any explicit reference, but is this acceptable coding standard?

refack replied:
IMHO what you did is great, that's exactly one of the reasons to avoid recycling.
Now you stop the function on first error and propagate it, and not continue to process in an error state.
Double win 🥇

refack replied:
BTW: it's was simpler to do this now because of the RAII semantics of ares_unique_ptr.
Triple win 🏆

refack commented:
> I've updated my PR with new changes as discussed, using std::unique_ptr instead of MallocedBuffer<>.

Thanks, IMHO that part looks nice and succinct 👍

refack commented:
CI: 23628

addaleax commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/18021/

uttampawar commented:
@refack @addaleax I've rebased this patch on current master, and made changes addressing all the concerns. Looking for any feedback you may have. 
Some of the test results are,

$ make -j test
...
[==========] 80 tests from 10 test cases ran. (194 ms total)
[  PASSED  ] 80 tests.
...
[01:58|% 100|+ 2442|-   0]: Done

$ git log $TRAVIS_COMMIT_RANGE --pretty=format:'%h' --no-merges | tail -1 | xargs npx core-validate-commit --no-validate-metadata
npx: installed 13 in 2.01s
  ✔  02aced6f21fbef6949a3a9ebdeb20454c4a99909
     ✔  0:0      skipping fixes-url                        fixes-url
     ✔  0:0      blank line after title                    line-after-title
     ✔  0:0      line-lengths are valid                    line-length
     ✔  0:0      valid subsystems                          subsystem
     ✔  0:0      Title is formatted correctly.             title-format
     ⚠  0:50     Title should be <= 50 columns.            title-length

addaleax commented:
@refack Could you take another look?

refack commented:
New CI: https://ci.nodejs.org/job/node-test-pull-request/18079/

refack commented:
Test failure probably not-related Refs: https://github.com/nodejs/node/issues/23807
[`debian9-64`](https://ci.nodejs.org/job/node-test-commit-linux/22605/nodes=debian9-64/testReport/junit/(root)/test/parallel_test_tls_set_secure_context/)
```
test.parallel/test-tls-set-secure-context
Error Message: timeout
```
[`fedora-latest-x64`](https://ci.nodejs.org/job/node-test-commit-linux/22605/nodes=fedora-latest-x64/testReport/junit/(root)/test/parallel_test_tls_set_secure_context/)
```
test.parallel/test-tls-set-secure-context
Error Message: timeout
```

refack commented:
> Do we need to re-run it?

Yes, here we go https://ci.nodejs.org/job/node-test-pull-request/18087/

Trott commented:
Thanks for the contribution! 🎉

(If you're interested in other possible contributions to Node.js but don't have a good idea of where to start looking, some ideas are posted at https://www.nodetodo.org/next-steps/.)

uttampawar commented:
@Trott @refack Thank you for your support. I'll definitely look into above site for more ideas but I'm more interested in performance related opportunities specially where we can leverage CPU or platform features.

MylesBorins commented:
Should this be backported to v8.x and v10.x or should it bake a bit more
